### ISSUE TITLE
JetBrains: Use IntelliJ SDK for showing code autocompletions

### ISSUE THREAD

**Issue by vdavid** [sourcegraph/sourcegraph#53935]
Newer versions of the IntelliJ SDK allows plugins to register inline completions.
We should use this API when running on a compatible IntelliJ version. Docs [here](https://github.com/JetBrains/intellij-community/tree/master/platform/platform-impl/src/com/intellij/codeInsight/inline/completion).

We'll still need to retain and maintain the manual version for older IDE versions, though, for a year or so.

**Comment by vdavid** [comment]
@Gedochao, some good news for you 🙂 Even if it's not an immediate help, but at least a brighter future.

**Comment by vdavid** [comment]
We suspect that Copilot is not using the official API. If that's the case then it might be because
1) it doesn’t allow them to collect the right telemetry
2) it doesn’t provide exactly the same UX that enables them to deliver the desired acceptance rate

Let's do a spike first and see if this suits us.

**Comment by Gedochao** [comment]
There's no way Copilot would be using the new official API as it's bleeding edge, only available in `2023.2 EAP`.
It's possible to play with it that way, but we cannot really use it in our code yet (and won't be able to for a long time, as I assume it will only be available on production starting with `2023.2` onwards.
And even then, the API is annotated as experimental, so I wonder if it's okay to rely on it just yet.

**Comment by Gedochao** [comment]
However, I do think we might get somewhere by trying to mimic the inside workings of the new API while implementing the completions ourselves... maybe.
if anyone wants to try it out, the version to build against to set in `Gradle` is `LATEST-EAP-SNAPSHOT`

**Comment by vdavid** [comment]
I'll close this as wontfix.

Copy-paste from Slack, written by @Gedochao:

"Just a very short update on the new completion API for JetBrains SDK... correct me if I'm wrong, but it seems that the API is literally only visible in the master branch of the repo.
The changes have been merged on :jet-brains:  branch for 232.6095.10, while the newest version available from maven is 231.9161.41
The APIs are also annotated as Experimental and looking at their git history, they have been evolving rapidly in the recent  months (there's changes from 2 days ago)
the new APIs don't seem to be available for any stable version of IntelliJ.
There might be a way to use a snapshot from somewhere or to build the sources ourselves to play with it, but I do not think it's worth the time given that it's very uncertain if the shape they're in is production-ready and how much of the current behaviours it would retain before release.
We could reach out to JetBrains and ask, but realistically speaking, I do not think it's something we'll be able to rely on anytime soon.
TL;DR at a glance, I do not think this is worth the time for now, we'll probably be stuck with the old APIs for a long while, since the new stuff is bleeding edge and still unreleased"

"I'd assume we'll see this in a mainline release somewhere in 2023.2, which should be soon.
but I assume the @Experimental annotation will probably stay there.
So it's probably a good idea to keep the spike ticket open and play with it later."

"update: managed to build a plugin against IntelliJ LATEST-EAP-SNAPSHOT
I'd still say everything I've written above holds true.
However, I think we might be able to take inspiration from the experimental API code while trying to implement the completions manually ourselves."

"the com.intellij.codeInsight.inline.completion is unavailable in 2023.1 , it's only in the 2023.2 EAP... and that's the stuff that was linked in the ticket.
It's possible it exists in some form somewhere else and I just didn't find it. Their codebase is a maze."

"Additionally, the APIs in com.intellij.codeInsight.inline.completion are all annotated as @ApiStatus.Experimental, which doesn't really make me trust it's stable :pepe-wipe:"

"I didn't really manage to use it, despite building against the EAP version, as the extension points for the plugin.xml didn't seem to be in place.
this being bleeding edge also means there's no examples anywhere to refer to that I found."

"it's not unlikely that there's more to find out somewhere. it just didn't feel like it's ready for production usage just yet"
